src/Engine/UnitActionLifecycles/LaunchLifecycle.java

```diff
@@ -0,0 +1,212 @@
+            cargo.y = actor.y;
+
+            ArrayList<XYCoord> destinations = Utils.findPossibleDestinations(cargo, map);
+            // Acting in place after a launch is a cool concept, but a little weird in execution
+            // Also, allowing action in place would allow *launching*, and recursive launching isn't something I wanna bite off
+            destinations.remove(moveLocation);
+
+            // Build a launch action for each possible action the cargo can do after launch
+            for( XYCoord coord : destinations )
+            {
+              Path cargoMovePath = Utils.findShortestPath(cargo, coord, map);
+              ArrayList<GameActionSet> cargoActions = cargo.getPossibleActions(map, cargoMovePath, ignoreResident);
+
+              for( GameActionSet actionSet : cargoActions )
+                for( GameAction action : actionSet.getGameActions() )
+                  launchActions.add(new LaunchAction(actor, cargo, action));
+            }
+
+            // ...and reset it when we're done calculating
+            cargo.x = -1;
+            cargo.y = -1;
+          }
+
+          if( !launchActions.isEmpty() )
+          {
+            return new GameActionSet(launchActions);
+          }
+        }
+      }
+      return null;
+    }
+
+    @Override
+    public String name()
+    {
+      return "LAUNCH";
+    }
+
+    /**
+     * From Serializable interface
+     * @return The statically-defined object to use for this action type.
+     */
+    private Object readResolve()
+    {
+      return LAUNCH;
+    }
+  } // ~Factory
+
+  public static class LaunchAction implements GameAction
+  {
+    private Unit launcher, launchee;
+    private GameAction cargoAction;
+
+    public LaunchAction(Unit actor, Unit passenger, GameAction passengerAction)
+    {
+      launcher = actor;
+      launchee = passenger;
+      cargoAction = passengerAction;
+    }
+
+    @Override
+    public GameEventQueue getEvents(MapMaster gameMap)
+    {
+      // UNLOAD actions consist of
+      //   MOVE (transport)
+      //   UNLOAD
+      //   [UNLOAD]*
+      GameEventQueue launchEvents = new GameEventQueue();
+
+      // Validate input.
+      boolean isValid = true;
+      isValid &= null != launchee && !launchee.isTurnOver;
+      isValid &= null != launcher && !launcher.isTurnOver && launcher.heldUnits.contains(launchee);
+      isValid &= null != cargoAction;
+      isValid &= null != gameMap;
+
+      // Generate events.
+      if( isValid )
+      {
+        launchEvents.add(new LaunchEvent(launcher, launchee));
+        GameEventQueue cargoEvents = cargoAction.getEvents(gameMap);
+        launchEvents.addAll(cargoEvents);
+
+        XYCoord firstEndPoint = cargoEvents.peekFirst().getEndPoint();
+        if (null != firstEndPoint && firstEndPoint.equals(launcher.x, launcher.y))
+          // If the first action ends on top of the launcher, assume it's a move action and re-load the unit to avoid weird states
```

Comment:
You explicitly disallow launching to the start point elsewhere - why special-case it here?

Reply:
I'd like to defend against the board getting in weird states. Say we move into fog and somehow get ambushed with no empty spaces to move into -- then this logic would be useful.
Also, I was *initially* planning to allow acting on top of the launcher, but then I realized there was no easy way to allow any action in place without letting you plan to e.g. attack from on top of a *different* unit.

Reply:
I might put that exceptional case in a comment here, just so it's immediately clear how that could happen.